ingester/ingester_lifecycle.go

```diff
@@ -0,0 +1,419 @@
+			ringDesc.AddIngester(i.id, i.addr, i.tokens, i.state)
+		} else {
+			ingesterDesc.Timestamp = time.Now().Unix()
+			ingesterDesc.State = i.state
+			ingesterDesc.Addr = i.addr
+			ringDesc.Ingesters[i.id] = ingesterDesc
+		}
+
+		return ringDesc, true, nil
+	})
+}
+
+// changeState updates consul with state transistions for us.  NB this must be
+// called from loop()!  Use ChangeState for calls from outside of loop().
+func (i *Ingester) changeState(state ring.IngesterState) error {
+	// Only the following state transitions can be triggered externally
+	if !((i.state == ring.PENDING && state == ring.JOINING) || // triggered by ClaimStart
+		(i.state == ring.PENDING && state == ring.ACTIVE) || // triggered by autoJoin
+		(i.state == ring.JOINING && state == ring.ACTIVE) || // triggered by ClaimFinish
+		(i.state == ring.ACTIVE && state == ring.LEAVING)) { // triggered by shutdown
+		return fmt.Errorf("Changing ingester state from %v -> %v is disallowed", i.state, state)
+	}
+
+	log.Infof("Changing ingester state from %v -> %v", i.state, state)
+	i.state = state
+	return i.updateConsul()
+}
+
+// transferChunks finds an ingester in PENDING state and transfers our chunks
+// to it.
+func (i *Ingester) transferChunks() error {
+	targetIngester := i.findTargetIngester()
+	if targetIngester == nil {
+		return fmt.Errorf("cannot find ingester to transfer chunks to")
+	}
+
+	log.Infof("Sending chunks to %v", targetIngester.Addr)
+	client, err := client.MakeIngesterClient(targetIngester.Addr, claimRemoteTimeout)
+	if err != nil {
+		return err
+	}
+	defer client.(io.Closer).Close()
+
+	ctx := user.Inject(context.Background(), "-1")
+	stream, err := client.TransferChunks(ctx)
+	if err != nil {
+		return err
+	}
+
+	for userID, state := range i.userStates.cp() {
+		for pair := range state.fpToSeries.iter() {
+			state.fpLocker.Lock(pair.fp)
+
+			chunks, err := toWireChunks(pair.series.chunkDescs)
+			if err != nil {
+				state.fpLocker.Unlock(pair.fp)
+				return err
+			}
+
+			err = stream.Send(&cortex.TimeSeriesChunk{
+				FromIngesterId: i.id,
+				UserId:         userID,
+				Labels:         util.ToLabelPairs(pair.series.metric),
+				Chunks:         chunks,
+			})
+			state.fpLocker.Unlock(pair.fp)
+			if err != nil {
+				return err
+			}
+
+			sentChunks.Add(float64(len(chunks)))
+		}
+	}
+
+	_, err = stream.CloseAndRecv()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// findTargetIngester finds an ingester in PENDING state.
+func (i *Ingester) findTargetIngester() *ring.IngesterDesc {
+	for j := 0; j < 10; j++ {
+		ringDesc, err := i.consul.Get(ring.ConsulKey)
+		if err != nil {
+			log.Errorf("Error talking to consul: %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		ingesters := ringDesc.(*ring.Desc).FindIngestersByState(ring.PENDING)
+		if len(ingesters) <= 0 {
+			log.Warnf("No pending ingesters found...")
+			time.Sleep(1 * time.Second)
+			continue
+		}
+
+		return ingesters[0]
+	}
```

Comment:
The retry values here (num retries, time between retries) should probably be flags, since we've got no idea what they should be.